updater/lib/github_api/dependency_submission.rb

```diff
@@ -88,70 +86,67 @@ def symbolic_ref
-                # Dependabot::Dependency objects do not include immediate dependencies,
-                # this is a capability each parser will need to have added.
-              ],
+              # We expect direct dependencies to be added to the metadata, but they may not always be available
+              dependencies: dep.metadata.fetch(:depends_on, []),
               metadata: {}
             }
           end
         }
-      end
+      }
     end
 
-    # For each distinct directory in our manifest list, we only want the highest priority manifests available,
-    # this method will filter out manifests for directories that have lockfiles and so on.
-    sig { params(snapshot: Dependabot::DependencySnapshot).returns(T::Array[Dependabot::DependencyFile]) }
-    def relevant_manifests(snapshot)
-      manifests_by_directory = snapshot.dependency_files.each_with_object({}) do |file, dirs|
-        # If the file doesn't have any dependencies assigned to it, then it isn't relevant.
-        next if file.dependencies.empty?
-
-        # Build up a dictionary of unique directories...
-        dirs[file.directory] ||= {}
-        # Add a list of files for each distinct priority...
-        dirs[file.directory][file.priority] ||= []
-        dirs[file.directory][file.priority] << file
+    # Dependabot aligns with Dependency Graph's existing behaviour where all dependencies are attributed to the
+    # most specific file out of the manifest or lockfile for the directory rather than split direct and indirect
+    # to the manifest and lockfile respectively.
+    #
+    # Dependabot's parsers apply this precedence by deterministic ordering, i.e. the manifest file's dependencies
+    # are added to the set first, then the lockfiles so we want the right-most file in the set, excluding anything
+    # marked as a support file.
+    sig { params(dependency_files: T::Array[Dependabot::DependencyFile]).returns(Dependabot::DependencyFile) }
+    def relevant_dependency_file(dependency_files)
+      filtered_files = dependency_files.reject { |f| f.support_file? || f.vendored_file? }
+
+      # TODO(brrygrdn): Make relevant_dependency_file an ecosystem property
+      #
+      # It turns out that the right-most-file-aligns-with-dependency-graph-static-parsing strategy isn't a durable
+      # assumption, for Go we prefer go.mod over go.sum even though the latter is technically the lockfile.
+      #
+      # With python ecosystems, this gets more fragmented and it isn't always accurate for 'bun' Javascript projects
+      # if they use mixins.
+      #
+      # The correct way to solve this is to use Dependency Injection to provide a small ecosystem-specific helper
+      # for this and PURLs so we can define the correct heuristic as necessary and use the 'last file wins' as our
+      # fallback strategy.
+      if %w(bun go Python).include?(package_manager)
+        T.must(filtered_files.first)
+      else
+        T.must(filtered_files.last)
```

Comment:
Unfortunately, my best effort strategy blew up early, so this is a small tactical change to just avoid the worst outcomes for now.

We still need to solve the scenario where the `dependencies` data we have at this point doesn't contain all the metadata we need to supply a fully populated SBOM for a project, and the way we scrap that metadata will be ecosystem specific too.

My thinking is that this class should become a generic/base class and each ecosystem should be expected to nominate its own subclass that:
- Implements this method to provide a heuristic to determine the relevant file out of a set for a single dir
- Implements a method to decorate the `dependencies` object with `metadata[:depends_on` if it isn't populated by the parser itself and requires extra native calls/processing

Reply:
This also allows us to reflect on an ecosystem module and ask it if it implements graphing or not so we can start to build up a contract and runtime validation that a new ecosystem has a 'grapher' as well as a 'parser'; it also pushes the testing contract out of the updater and into the specific ecosystems as a nice workflow bonus.

Reply:
First draft of this concept here:
- https://github.com/dependabot/dependabot-core/pull/12998